src/go/guestagent/pkg/iptables/iptables.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright © 2024 SUSE LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package iptables handles forwarding ports found in iptables DNAT
+package iptables
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Masterminds/log-go"
+	"github.com/docker/go-connections/nat"
+	"github.com/lima-vm/lima/pkg/guestagent/iptables"
+	"github.com/rancher-sandbox/rancher-desktop/src/go/guestagent/pkg/tracker"
+)
+
+// ForwardPorts forwards ports found in iptables DNAT. In some environments,
+// like WSL, ports defined using the CNI portmap plugin happen through iptables.
+// These ports are not sent to places like /proc/net/tcp and are not picked up
+// as part of the normal forwarding system. This function detects those ports
+// and binds them so that they are picked up.
+// The argument is a time, in seconds, to wait between updating.
+func ForwardPorts(ctx context.Context, tracker tracker.Tracker, updateInterval time.Duration) error {
+	var ports []iptables.Entry
+	for {
+		// Detect ports for forward
+		newPorts, err := iptables.GetPorts()
+		if err != nil {
+			// iptables exiting with an exit status of 4 means there
+			// is a resource problem. For example, something else is
+			// running iptables. In that case, we can skip trying it for
+			// this loop. You can find the exit code in the iptables
+			// source at https://git.netfilter.org/iptables/tree/include/xtables.h
+			if strings.Contains(err.Error(), "exit status 4") {
+				log.Debug("iptables exited with status 4 (resource error). Retrying...")
+				time.Sleep(updateInterval)
+				continue
+			}
+			return err
+		}
+		log.Debugf("iptable found ports %+v", newPorts)
+		// Diff from existing forwarded ports
+		added, removed := comparePorts(ports, newPorts)
+		ports = newPorts
+		// Remove old forwards
+		for _, p := range removed {
+			name := entryToString(p)
+			if err := tracker.Remove(generateID(name)); err != nil {
+				log.Warnf("failed to close listener %q: %w", name, err)
+			}
+		}
+		portMap := make(nat.PortMap)
+		// Add new forwards
+		for _, p := range added {
+			if p.TCP {
```

Comment:
Why only TCP? I thought we also do UDP, or am I confused?

Reply:
We totally support UDP now in RD but, this decision came from how the `iptables.Entry` defined in the [upstream](https://github.com/lima-vm/lima/blob/adc5a71943b72949aefcc2b9eb8ffb1f7fe2dce5/pkg/guestagent/iptables/iptables.go#L14). It doesn't seem to give the proper protocol but rather a boolean. Of course, if `TCP` is set to true, we know it is 100% `TCP` but when it's set to false I'm not sure if that means it would be UDP 100%.